scipy/sparse/_dok.py

```diff
@@ -207,63 +310,43 @@ def __add__(self, other):
         if isscalarlike(other):
             res_dtype = upcast_scalar(self.dtype, other)
             new = self._dok_container(self.shape, dtype=res_dtype)
-            # Add this scalar to every element.
-            M, N = self.shape
-            for key in itertools.product(range(M), range(N)):
+            # Add this scalar to each element.
+            for key in itertools.product(*[range(d) for d in self.shape]):
                 aij = self._dict.get(key, 0) + other
                 if aij:
                     new[key] = aij
-            # new.dtype.char = self.dtype.char
         elif issparse(other):
+            if other.shape != self.shape:
+                raise ValueError("Matrix dimensions are not equal.")
+            res_dtype = upcast(self.dtype, other.dtype)
+            new = self._dok_container(self.shape, dtype=res_dtype)
+            new._dict = self._dict.copy()
             if other.format == "dok":
-                if other.shape != self.shape:
-                    raise ValueError("Matrix dimensions are not equal.")
-                # We could alternatively set the dimensions to the largest of
-                # the two matrices to be summed.  Would this be a good idea?
-                res_dtype = upcast(self.dtype, other.dtype)
-                new = self._dok_container(self.shape, dtype=res_dtype)
-                new._dict.update(self._dict)
                 with np.errstate(over='ignore'):
-                    new._dict.update((k, new[k] + other[k]) for k in other.keys())
+                    new._dict.update((k, new[k] + v) for k, v in other.items())
```

Comment:
Let's unify this with the version below:

```python
if other.format == "dok":
  o_items = other.items()
else:
  other = other.tocoo()
  if self.ndim == 1:
    o_items = zip(other.coords[0], other.data)
  else:
    o_items = zip(zip(*other.coords), other.data)
```

Reply:
Nice!